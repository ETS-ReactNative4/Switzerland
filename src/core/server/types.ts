export type Args = { node: HTMLElement };

