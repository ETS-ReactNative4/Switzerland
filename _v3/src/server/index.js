/**
 * @function render ∷ ∀ a. [String, [(p → p)]]|String → Object String a → String
 */
export async function render(component, props = {}) {
    const node = await component.render(props);
    return node.outerHTML;
}
