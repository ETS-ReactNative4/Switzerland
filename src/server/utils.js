/**
 * @function initialProps ∷ HTMLElement e, Props p ⇒ e → [(p → p)] → p
 */
export const initialProps = (node, middleware) => {
    const utils = {
        dispatch: () => {},
        abort: () => {},
        getMiddleware: () => Promise.all(middleware),
        getLatestProps: () => undefined,
        isHeadless: true,
        isIdle: () => false,
        isResolved: () => true
    };

    return {
        signal: {},
        node,
        utils,
        render: () => {},
        prevProps: undefined,
        lifecycle: 'mount'
    };
};
