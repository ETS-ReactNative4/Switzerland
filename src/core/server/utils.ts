import { Child, Node, Element } from '../../types';
import { getWindow } from '../../utils';
import * as server from '../server';

export async function transform(parentNode: HTMLElement, child: Child): Promise<string> {
    const window = await getWindow();

    if (typeof child === 'string') {
        const node = window.document.createTextNode(child);
        parentNode.appendChild(node);
        return parentNode.outerHTML;
    }

    if (isNode(child)) {
        if (isHTMLElement(child.element)) {
            const node = window.document.createElement(child.element);
            parentNode.appendChild(node);

            if (child.children != null)
                await Promise.all(child.children.map(async (child) => await transform(node, child)));

            return node.outerHTML;
        }

        if (isSwiss(child.element)) {
            const node = await child.element.render(child.attributes);
            parentNode.appendChild(node);

            if (child.children != null)
                await Promise.all(child.children.map(async (child) => await transform(node, child)));

            return node.outerHTML;
        }
    }

    return '';
}

function isNode(child: Child): child is Node {
    return 'element' in Object(child);
}

function isHTMLElement(element: Element): element is string {
    return typeof element === 'string';
}

function isSwiss(element: Element): element is server.Swiss {
    return element instanceof server.Swiss;
}
