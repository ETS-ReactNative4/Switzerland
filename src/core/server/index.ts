import { DOMWindow } from 'jsdom';
import * as abstract from '../abstract';
import { getWindow } from '../../utils';
import type { Attributes, View } from '../../types';
import { transform } from './utils';

export class Swiss extends abstract.Swiss {
    protected name: string;
    protected extend: null | string;
    protected view: View;

    constructor(name: string, extend: null | string = null, view: View) {
        super();
        this.name = name;
        this.extend = extend;
        this.view = view;
    }

    async render(attributes: Attributes = {}): Promise<HTMLElement> {
        // Initialte the node and defined the `data-swiss` attribute which denotes a boundary.
        const window = (await getWindow()) as DOMWindow;
        const node = window.document.createElement(this.name);
        node.setAttribute('data-swiss', '');

        // Define the optional element to be extended from which will inherit its behaviours.
        if (this.extend != null) node.setAttribute('is', this.extend);

        // Define all of the attributes for the host node.
        for (const [key, value] of Object.entries(attributes)) node.setAttribute(key, value);

        // Render the current view and all of its descendant views.
        await transform(node, this.view());

        return node;
    }
}

export function create(
    name: string,
    extend: null | string = null,
    view: View = () => null
): InstanceType<typeof Swiss> {
    return new Swiss(name, extend, view);
}
