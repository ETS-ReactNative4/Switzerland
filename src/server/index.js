import browserEnv from 'browser-env';
import { elements } from '../core/index.js';
import * as u from '../core/utils.js';
import { initialProps } from './utils.js';

browserEnv(['window', 'document', 'navigator']);

export async function render([name, middleware], attrs = {}, node = document.createElement(name)) {
    // Attach any custom attributes to the host element.
    Object.entries(attrs).forEach(([name, value]) => node.setAttribute(name, value));

    // Cycle through the middleware and append the "resolved" class name.
    const props = initialProps(node, middleware);
    await u.cycleMiddleware(node, props, middleware);
    node.classList.add('resolved');

    // Render recursively each custom child element.
    for (const name of elements.keys()) {
        const childNode = node.querySelector(name);
        const middleware = elements.get(name);
        childNode && (await render([name, middleware], {}, childNode));
    }

    return node.outerHTML;
}
