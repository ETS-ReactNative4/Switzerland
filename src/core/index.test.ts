import { create, h, render } from '.';
import * as server from './server';

describe('create', () => {
    it('should be able to create component instances', () => {
        expect(create('x-maria')).toBeInstanceOf(server.Swiss);
    });
});

describe('render', () => {
    it('should be able to render component hierarchies', async () => {
        const adam = create('x-adam', () => 'Adam');
        const maria = create('x-maria', () => 'Maria');
        const imogen = create('x-imogen', () => 'Imogen');

        const people = create('x-people', () => {
            return h('section', {}, [
                h(adam, { nationality: 'British' }),
                h(maria, { nationality: 'Russian/British' }),
                h(imogen, { nationality: 'British' }),
            ]);
        });

        const html = await render(people);
        expect(html).toMatchSnapshot();
    });
});
