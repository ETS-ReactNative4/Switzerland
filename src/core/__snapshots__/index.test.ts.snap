// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should be able to render component hierarchies 1`] = `"<x-people data-swiss=\\"\\"><section><x-adam data-swiss=\\"\\" nationality=\\"British\\">Adam</x-adam><x-maria data-swiss=\\"\\" nationality=\\"Russian/British\\">Maria</x-maria><x-imogen data-swiss=\\"\\" nationality=\\"British\\">Imogen</x-imogen></section></x-people>"`;
